### PR TITLE
Expose data on api

### LINKED ISSUE #21: Expose data on API

### PR DESCRIPTION
**Pull request by baltazarix**
expose data from models `Case`, `Story` at `/data.json`

closes #21

### CODE REVIEW

File: victims/models.py [hunk 1]
```diff
@@ -27,6 +29,20 @@ def is_valid(self):
 
         return True
 
+    def to_JSON(self):
```
Comment (cuducos) [inline victims/models.py]: `to_JSON` is a syntax that [does not follow PEP8](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables):

> **Method Names and Instance Variables**
> 
> Use the function naming rules: lowercase with words separated by underscores as necessary to improve readability.

Can we rename it to `to_json`?

File: victims/models.py [hunk 2]
```diff
@@ -27,6 +29,20 @@ def is_valid(self):
...
+        """
+        Serialize object to json
+        """
```
Comment (cuducos) [inline victims/models.py]: This short docstring could be in one:

```
def to_json(self):
    """Serialize object to json"""
```

But is this docstring really necessary? I feel like the name `to_json` says it all and the docstring is redundant. What do you think about it?

File: victims/models.py [hunk 3]
```diff
@@ -27,6 +29,20 @@ def is_valid(self):
...
+        d = {
```
Comment (cuducos) [inline victims/models.py]: What does `d` stands for? May we have a more meaningful variable name here?

File: victims/models.py [hunk 4]
```diff
@@ -48,3 +64,18 @@ def tags_and_colors(self):
...
+            "id": self.id,
+            "aggressor_side": self.aggressor_side,
+            "when": str(self.when),
+            "state": self.state,
+            "city": self.city,
+            "tags": self.tags,
+            "stories": [story.to_JSON() for story in self.stories],
+        }
+        return json.dumps(d, sort_keys=True)
```
Comment (cuducos) [inline victims/models.py]: Just the same as my comments in the other `dataclass` ; )

**Review by cuducos — CHANGES_REQUESTED** [review]

**Comment by baltazarix** [comment]
@cuducos, Just pushed changes according to your comments. Thanks for time you've spent on this review and also for right remarks.

File: victims/models.py [hunk 5]
```diff
@@ -27,6 +29,17 @@ def is_valid(self):
...
 
+    def to_json(self):
+        items = {
+            "url": self.url,
+            "source": self.source,
+            "title": self.title,
+            "image_or_video": self.image_or_video,
+            "summary": self.summary,
+            "case_id": self.case_id,
+        }
```
Comment (cuducos) [inline victims/models.py]: Any special reason to not use `dict(self)` and keep the descriptive verbose approach? IMHO this is technical debt.

File: victims/models.py [hunk 6]
```diff
@@ -48,3 +61,15 @@ def tags_and_colors(self):
...
+    def to_json(self):
+        items = {
+            "id": self.id,
+            "aggressor_side": self.aggressor_side,
+            "when": str(self.when),
+            "state": self.state,
+            "city": self.city,
+            "tags": self.tags,
+            "stories": [story.to_JSON() for story in self.stories],
+        }
```
Comment (cuducos) [inline victims/models.py]: Any special reason to not use `dict(self)` and keep the descriptive verbose approach? IMHO this is technical debt.

File: victims/__init__.py [hunk 7]
```diff
@@ -47,3 +47,10 @@ def force_https(request):
 @jinja.template("about.html")
 async def about(request):
     return {"title": TITLE, "url_path": "/about.html"}
+
+
+@app.route("/data.json")
+async def data(request):
+    cases = await get_cases()
+    cases = [case.to_JSON() for case in cases]
```
Comment (cuducos) [inline victims/__init__.py]: There's no `case.to_JSON()` anymore, right?

**Comment by cuducos** [comment]
> Just pushed changes according to your comments.

I think you missed some points. Could you review or come back with questions to help me clarify my points? So sorry if I was unable to make myself clearer : (

File: victims/models.py [hunk 8]
```diff
@@ -48,3 +61,15 @@ def tags_and_colors(self):
...
+
+    def to_json(self):
+        items = {
+            "id": self.id,
+            "aggressor_side": self.aggressor_side,
+            "when": str(self.when),
+            "state": self.state,
+            "city": self.city,
+            "tags": self.tags,
+            "stories": [story.to_JSON() for story in self.stories],
```
Comment (turicas) [inline victims/models.py]: This line should not encode `story` as JSON since the whole `items` is encoded in the end.

**Review by turicas — CHANGES_REQUESTED** [review]

File: victims/models.py (hunk 8 above)
Comment (baltazarix) [reply victims/models.py]: good point, thanks

File: victims/__init__.py (hunk 7 above)
Comment (baltazarix) [reply victims/__init__.py]: you're right, my bad

File: victims/models.py (hunk 5 above)
Comment (baltazarix) [reply victims/models.py]: I used variable here cause I had to somehow serialize `datetime.date`. After my research on network I can propose to implement custom json encoder class, which will handle `datetime.date` value. With usage of `dataclasses.asdict`. Then whole `to_json` method may look like this:

```python
    def to_json(self):
        return json.dumps(asdict(self), cls=DateEncoder, sort_keys=True)
```

File: victims/models.py (hunk 8 above)
Comment (turicas) [reply victims/models.py]: @baltazarix do you mind adding a simple test for this endpoint? [It's pretty simple](https://sanic.readthedocs.io/en/latest/sanic/testing.html) and can avoid assure we're serializing data correctly.

File: victims/models.py (hunk 5 above)
Comment (cuducos) [reply victims/models.py]: This is a good way. Alternatively simpler though:

```python
def to_json(self):
    data = asdict(self)
    data['when'] = date['when'].strftime("%Y-%m-%d")
    return json.dumps(data, sort_keys=True)
```

File: victims/models.py (hunk 8 above)
Comment (baltazarix) [reply victims/models.py]: @turicas, sure! I was thinking about it but I've had no idea how test scenario should look like. Only ideas I have are:
- checking for status code
- create some test data (e.g. 1 `Case` instance, 2 `Story` instances), then compare response content.

Feel free to tell me if I'm missing something :)

Comment (baltazarix) [reply victims/models.py]: > Take a look at existing `test_home_contents`: we test some contents that we expect to see in the response. Name of the stories, maybe dates formatted as str (as they are the bottleneck of this serialization) etc…

Yep, I wanted to make something similar, sorry if I wasn't clear enough.

**Pull request closed without merging**